### ISSUE TITLE
KOReader not starting (Kindle Scribe)

### ISSUE THREAD

**Issue by swarzarossa** [koreader/koreader#11392]
* KOReader version:  v2024.01-42-g7f33d94c4_2024-01-19
* Device: ~~Kindle Oasis~~ Kindle Scribe

#### Issue
KOReader is not starting after updating to the 2024.01 release. Attempts to run the nightly version also resulted in failure, as indicated in the crash log. Also tried a clean install.

##### `crash.log`

```
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 01/19/24-23:22:14
 [*] Version: v2024.01-42-g7f33d94c4_2024-01-19

ffi.load: libs/libutf8proc.so.3
ffi.load: blitbuffer
ffi.load: sqlite3
01/19/24-23:22:15 INFO  initializing for device KindleScribe 
01/19/24-23:22:15 INFO  framebuffer resolution: {
  h = 2480,
  w = 1860
} --[[table: 0xb6d0b670]] 
./luajit: frontend/device/kindle/device.lua:1433: Error opening input device </dev/input/by-path/platform-11007000.i2c-event-joystick>: No such file or directory
stack traceback:
	[C]: in function 'open'
	frontend/device/kindle/device.lua:1433: in function 'init'
	frontend/device.lua:62: in main chunk
	[C]: in function 'require'
	./reader.lua:151: in main chunk
	[C]: at 0x00013ebd

```
</details>

**Comment by Frenzie** [comment]
Seeing how that relates to https://github.com/koreader/koreader/pull/11047 and #11159, pinging @NiLuJe and @mergen3107

**Comment by mergen3107** [comment]
Yeah, I've seen this issue from a user on 4PDA.

For now just comment out this line that tries to open the gyroscope. 

What happened last time was that this link in by-path folder is… missing. 

Please provide here the tree output of /dev/input/ folder with its subdirectories.

**Comment by mergen3107** [comment]
I'll test it, most probably we will end up just pointing to the actual eventX device, the same way we did with touch recently.

**Comment by NiLuJe** [comment]
I suspect it won't be stable (by which I mean, it won't always be enumerated at the same index), which is a problem ;).

**Comment by mergen3107** [comment]
Ok, then these complex selection loops then…

**Comment by mergen3107** [comment]
Sorry, the tweak might be dalayed. I just don't have enough time to play around with Kindle

**Comment by mergen3107** [comment]
> For now just comment out this line that tries to open the gyroscope.

@ERZUSO 
Can you please try adding this chunk of the code below this uncommented line?
```
    local std_out = io.popen("grep -A4 'accel' /proc/bus/input/devices | grep -o 'event[0-9]'", "r")
    if std_out then
        local gyro_dev = std_out:read("*line")
        std_out:close()
        if gyro_dev then
            self.input.open("/dev/input/"..gyro_dev)
        end
    end
```

This routine looks for the `eventX`, where `X` is the device number associated with the accelerometer.

@NiLuJe 
Do you see any problems with this approach?
Previously you mentioned that this is an overkill, but it seems we need to do this anyway...

**Comment by mergen3107** [comment]
From 4PDA forum feedback, issue seems to be complicated by different gyro device used. Can you please provide the contents of the file /proc/bus/input/devices ?

**Comment by NiLuJe** [comment]
Eh, that should do the trick for now, yeah ;).

**Comment by mergen3107** [comment]
4PDA forum user reported that /dev/input/event2 worked, but not event1. This is is the second one among these:
```

I: Bus=0018 Vendor=0000 Product=0000 Version=0000
N: Name="bma4xy_acc"
P: Phys=
S: Sysfs=/devices/virtual/input/input1
U: Uniq=
H: Handlers=event1 perfmgr 
B: PROP=0
B: EV=19
B: ABS=100 7
B: MSC=4

I: Bus=0018 Vendor=0000 Product=0000 Version=0000
N: Name="bma4xy_feature"
P: Phys=
S: Sysfs=/devices/virtual/input/input2
U: Uniq=
H: Handlers=event2 perfmgr 
B: PROP=0
B: EV=19
B: ABS=1000000
B: MSC=2
```

**Comment by NiLuJe** [comment]
Huh. What does `evtest` say about the detailed supported caps on each of those? (e.g., output from `evtest /dev/input/event1` & event2.

**Comment by mergen3107** [comment]
I asked on 4PDA, will test on mine later

**Comment by mergen3107** [comment]
@NiLuJe 
4PDA forum user responded, event2's evtest yields:
```

[kterm]# 
[kterm]# evtest /dev/input/event2
Input driver version is 1.0.1
Input device ID: bus 0x18 vendor 0x0 product 0x0 version 0x0
Input device name: "bma4xy_feature"
Supported events:
Event type 0 (Sync)
Event type 3 (Absolute)
Event code 24 (Pressure)
Value 19
Min 15
Max 22
Event type 4 (Misc)
Event code 1 (Pulseled)
Testing ... (interrupt to exit)
Event: time 1707153748.160808, type 3 (Absolute), code 24 (Pressure), value 15
Event: time 1707153748.160808, type 4 (Misc), code 1 (Pulseled), value 2
Event: time 1707153748.160808, -------------- Report Sync ------------
^[OPEvent: time 1707153749.594638, type 3 (Absolute), code 24 (Pressure), value 18
^[OREvent: time 1707153749.594638, type 4 (Misc), code 1 (Pulseled), value 2
Event: time 1707153749.594638, -------------- Report Sync ------------
Event: time 1707153750.680675, type 3 (Absolute), code 24 (Pressure), value 16
Event: time 1707153750.680675, type 4 (Misc), code 1 (Pulseled), value 2
Event: time 1707153750.680675, -------------- Report Sync ------------
^[OQEvent: time 1707153752.052785, type 3 (Absolute), code 24 (Pressure), value 17
Event: time 1707153752.052785, type 4 (Misc), code 1 (Pulseled), value 2
Event: time 1707153752.052785, -------------- Report Sync ------------
^[OSEvent: time 1707153752.872615, type 3 (Absolute), code 24 (Pressure), value 15
Event: time 1707153752.872615, type 4 (Misc), code 1 (Pulseled), value 2
Event: time 1707153752.872615, -------------- Report Sync ------------
^[OP^C
[kterm]# exit
```

Whereas event1 shows some weird characters like:
```

[kterm]# evtest /dev/input/event1
Input driver version is 1.0.1
Input device ID: bus 0x18 vendor 0x0 product 0x0 version 0x0
Input device name: "bma4xy_acc"
Supported events:
Event type 0 (Sync)
Event type 3 (Absolute)
Event code 0 (X)
Value 0
Min -2147483648
Max 2147483647
Event code 1 (Y)
Value 0
Min -2147483648
Max 2147483647
Event code 2 (Z)
Value 0
Min -2147483648
Max 2147483647
Event code 40 (Misc)
Value 0
Min 0
Max 0
Event type 4 (Misc)
Event code 2 (Gesture)
Testing ... (interrupt to exit)
^[OR^[OQ^[OR^[OP^[OS^[OQ^[OS^[OP^[OP^C
[kterm]#
```

**Comment by mergen3107** [comment]
These are while rotating Scribe around

**Comment by NiLuJe** [comment]
Okay, so, checking for [EV_ABS:ABS_PRESSURE](https://github.com/koreader/koreader/blob/962477e7c679c2724c369536d2a1e941c678ad1d/frontend/device/kindle/device.lua#L891) in the caps should do the trick...

I jkust have to actually cxobble that together in an input-tools modules, from the bits that are already in place in the external keyboard plugin...

**Comment by mergen3107** [comment]
Thanks :D pointing manually to event2 for this issue (and event1) directly works OK for now, I guess...

**Comment by Swatsev** [comment]
Help please. I have pw1. After waking up from sleep mode, a book page opens mixed with the home screen.

**Comment by mergen3107** [comment]
@Swatsev 
Can you please update to the latest nightly, and create a new ticket with verbose logs attached? 
Thanks!

**Comment by myrena** [comment]
> For now just comment out this line that tries to open the gyroscope.
Please tell me how to comment out the entry, in which file and on which line.

**Comment by mergen3107** [comment]
It is provided in the log:

frontend/device/kindle/device.lua:1433

**Comment by mergen3107** [comment]
> I just have to actually cobble that together in an input-tools module, from the bits that are already in place in the external keyboard plugin...

@NiLuJe 
Anything I can do to help you with this task please?

**Comment by NiLuJe** [comment]
Not really, but more data about the capabilities of *all* the input devices (even those we skip) will be useful once the framework is in place, so you can post the output of `/mnt/us/usbnet/bin/evtest` on *all* the input devices ;).

(No need to do anything, I'm just interested in the cap dump printed on startup after having chosen an input device).

**Comment by mergen3107** [comment]
Like this? I went through all of them one by one (while KOReader is up):
[evtest_results1.txt](https://github.com/koreader/koreader/files/14530294/evtest_results1.txt)

If yes, I then will ask on 4PDA as well.

**Comment by NiLuJe** [comment]
Yup, exactly :).

**Comment by mergen3107** [comment]
@NiLuJe 
Here is the evtest results from all /dev/input/ devices on one of the problematic devices (same as in OP). The user on 4PDA doesn't have usbnet installed, so they did it in kterm. Please let me know if you can see the file in this link. I am on mobile now, but I can attach the downloaded file itself later from PC if needed.

https://ds-blobs-2.cdn.devapps.ru/30307877/evtest.txt?s=000e282e0000000065edbf80000000006be9d9ba3501d654157a24bc82c574ef

**Comment by NiLuJe** [comment]
```
Script started on Sun 10 Mar 2024 02:36:34 PM GMT+3:0100
[kterm]# 
[kterm]# script evtest[J
[kterm]# evtest /dev/input/[J
by-path/  event1    event3    event5    touch
event0    event2    event4    stylus    uinput

[kterm]# evtest /dev/input/[Je
[kterm]# evtest /dev/input/event[J1[J0
Input driver version is 1.0.1
Input device ID: bus 0x19 vendor 0x1 product 0x1 version 0x100
Input device name: "bd71828-pwrkey"
Supported events:
  Event type 0 (Sync)
  Event type 1 (Key)
    Event code 116 (Power)
Testing ... (interrupt to exit)
^C
[kterm]# 
[kterm]# evtest /dev/input/event0[J[J1
Input driver version is 1.0.1
Input device ID: bus 0x18 vendor 0x0 product 0x0 version 0x0
Input device name: "bma4xy_acc"
Supported events:
  Event type 0 (Sync)
  Event type 3 (Absolute)
    Event code 0 (X)
      Value      0
      Min   -2147483648
      Max   2147483647
    Event code 1 (Y)
      Value      0
      Min   -2147483648
      Max   2147483647
    Event code 2 (Z)
      Value      0
      Min   -2147483648
      Max   2147483647
    Event code 40 (Misc)
      Value      0
      Min        0
      Max        0
  Event type 4 (Misc)
    Event code 2 (Gesture)
Testing ... (interrupt to exit)
^C
[kterm]# 
[kterm]# evtest /dev/input/event1[J[J2
Input driver version is 1.0.1
Input device ID: bus 0x18 vendor 0x0 product 0x0 version 0x0
Input device name: "bma4xy_feature"
Supported events:
  Event type 0 (Sync)
  Event type 3 (Absolute)
    Event code 24 (Pressure)
      Value     17
      Min       15
      Max       22
  Event type 4 (Misc)
    Event code 1 (Pulseled)
Testing ... (interrupt to exit)
^C
[kterm]# 
[kterm]# evtest /dev/input/event2[J[J3
Input driver version is 1.0.1
Input device ID: bus 0x18 vendor 0x2d1f product 0x158 version 0x1827
Input device name: "WacomDigitizer"
Supported events:
  Event type 0 (Sync)
  Event type 1 (Key)
    Event code 320 (ToolPen)
    Event code 321 (ToolRubber)
    Event code 330 (Touch)
    Event code 331 (Stylus)
    Event code 332 (Stylus2)
  Event type 3 (Absolute)
    Event code 0 (X)
      Value  15624
      Min        0
      Max    15624
    Event code 1 (Y)
      Value  20832
      Min        0
      Max    20832
    Event code 24 (Pressure)
      Value      0
      Min        0
      Max     4095
    Event code 25 (Distance)
      Value      0
      Min        0
      Max      255
    Event code 26 (XTilt)
      Value      0
      Min    -9000
      Max     9000
    Event code 27 (YTilt)
      Value      0
      Min    -9000
      Max     9000
Testing ... (interrupt to exit)
^C
[kterm]# 
[kterm]# evtest /dev/input/event3[J[J4
Input driver version is 1.0.1
Input device ID: bus 0x0 vendor 0x0 product 0x0 version 0x0
Input device name: "pt_mt"
Supported events:
  Event type 0 (Sync)
  Event type 1 (Key)
  Event type 2 (Relative)
  Event type 3 (Absolute)
    Event code 47 (MT Slot)
      Value      0
      Min        0
      Max        1
    Event code 48 (MT Major Axis Ellipse)
      Value      0
      Min        0
      Max      255
    Event code 53 (MT X)
      Value      0
      Min        0
      Max     1859
    Event code 54 (MT Y)
      Value      0
      Min        0
      Max     2479
    Event code 55 (MT Tool type)
      Value      0
      Min        0
      Max        2
    Event code 57 (MT Tracking ID)
      Value      0
      Min        0
      Max    65535
    Event code 58 (Pressure on contact area)
      Value      0
      Min        0
      Max      255
    Event code 59 (Contact hover distance)
      Value      0
      Min        0
      Max      255
Testing ... (interrupt to exit)
Event: time 1710070713.577541, type 3 (Absolute), code 57 (MT Tracking ID), value -1
Event: time 1710070713.577541, -------------- Report Sync ------------
Event: time 1710070716.381622, type 3 (Absolute), code 57 (MT Tracking ID), value 0
Event: time 1710070716.381622, type 3 (Absolute), code 53 (MT X), value 66
Event: time 1710070716.381622, type 3 (Absolute), code 54 (MT Y), value 2400
Event: time 1710070716.381622, type 3 (Absolute), code 58 (Pressure on contact area), value 74
Event: time 1710070716.381622, type 3 (Absolute), code 48 (MT Major Axis Ellipse), value 71
Event: time 1710070716.381622, -------------- Report Sync ------------
Event: time 1710070716.404068, type 3 (Absolute), code 57 (MT Tracking ID), value -1
Event: time 1710070716.404068, -------------- Report Sync ------------
Event: time 1710070717.833792, type 3 (Absolute), code 57 (MT Tracking ID), value 0
Event: time 1710070717.833792, type 3 (Absolute), code 53 (MT X), value 617
Event: time 1710070717.833792, type 3 (Absolute), code 54 (MT Y), value 2215
Event: time 1710070717.833792, type 3 (Absolute), code 58 (Pressure on contact area), value 83
Event: time 1710070717.833792, type 3 (Absolute), code 48 (MT Major Axis Ellipse), value 95
Event: time 1710070717.833792, -------------- Report Sync ------------
^C
[kterm]# 
[kterm]# evtest /dev/input/event4[J[J5
Input driver version is 1.0.1
Input device ID: bus 0x18 vendor 0x2d1f product 0x158 version 0x1827
Input device name: "stylus-custom"
Supported events:
  Event type 0 (Sync)
  Event type 1 (Key)
    Event code 320 (ToolPen)
    Event code 321 (ToolRubber)
    Event code 330 (Touch)
    Event code 331 (Stylus)
    Event code 332 (Stylus2)
  Event type 3 (Absolute)
    Event code 0 (X)
      Value      0
      Min        0
      Max     1860
    Event code 1 (Y)
      Value      0
      Min        0
      Max     2480
    Event code 24 (Pressure)
      Value      0
      Min        0
      Max     4095
    Event code 25 (Distance)
      Value      0
      Min        0
      Max      255
    Event code 26 (XTilt)
      Value      0
      Min    -9000
      Max     9000
    Event code 27 (YTilt)
      Value      0
      Min    -9000
      Max     9000
Testing ... (interrupt to exit)
^C
[kterm]# 
[kterm]# evtest /dev/input/event5[J[J[J[J[J[J[J
by-path/  event1    event3    event5    touch
event0    event2    event4    stylus    uinput

[kterm]# evtest /dev/input/[Jst
[kterm]# evtest /dev/input/stylus [J
Input driver version is 1.0.1
Input device ID: bus 0x18 vendor 0x2d1f product 0x158 version 0x1827
Input device name: "WacomDigitizer"
Supported events:
  Event type 0 (Sync)
  Event type 1 (Key)
    Event code 320 (ToolPen)
    Event code 321 (ToolRubber)
    Event code 330 (Touch)
    Event code 331 (Stylus)
    Event code 332 (Stylus2)
  Event type 3 (Absolute)
    Event code 0 (X)
      Value  15624
      Min        0
      Max    15624
    Event code 1 (Y)
      Value  20832
      Min        0
      Max    20832
    Event code 24 (Pressure)
      Value      0
      Min        0
      Max     4095
    Event code 25 (Distance)
      Value      0
      Min        0
      Max      255
    Event code 26 (XTilt)
      Value      0
      Min    -9000
      Max     9000
    Event code 27 (YTilt)
      Value      0
      Min    -9000
      Max     9000
Testing ... (interrupt to exit)
^C
[kterm]# 
[kterm]# evtest /dev/input/stylus [J[J[J[J[J[J[J[Jto
[kterm]# evtest /dev/input/touch [J
Input driver version is 1.0.1
Input device ID: bus 0x0 vendor 0x0 product 0x0 version 0x0
Input device name: "pt_mt"
Supported events:
  Event type 0 (Sync)
  Event type 1 (Key)
  Event type 2 (Relative)
  Event type 3 (Absolute)
    Event code 47 (MT Slot)
      Value      0
      Min        0
      Max        1
    Event code 48 (MT Major Axis Ellipse)
      Value      0
      Min        0
      Max      255
    Event code 53 (MT X)
      Value      0
      Min        0
      Max     1859
    Event code 54 (MT Y)
      Value      0
      Min        0
      Max     2479
    Event code 55 (MT Tool type)
      Value      0
      Min        0
      Max        2
    Event code 57 (MT Tracking ID)
      Value      0
      Min        0
      Max    65535
    Event code 58 (Pressure on contact area)
      Value      0
      Min        0
      Max      255
    Event code 59 (Contact hover distance)
      Value      0
      Min        0
      Max      255
Testing ... (interrupt to exit)
Event: time 1710070748.174057, type 3 (Absolute), code 57 (MT Tracking ID), value -1
Event: time 1710070748.174057, -------------- Report Sync ------------
Event: time 1710070750.715248, type 3 (Absolute), code 57 (MT Tracking ID), value 0
Event: time 1710070750.715248, type 3 (Absolute), code 53 (MT X), value 64
Event: time 1710070750.715248, type 3 (Absolute), code 54 (MT Y), value 2406
Event: time 1710070750.715248, type 3 (Absolute), code 58 (Pressure on contact area), value 80
Event: time 1710070750.715248, -------------- Report Sync ------------
Event: time 1710070750.762699, type 3 (Absolute), code 57 (MT Tracking ID), value -1
Event: time 1710070750.762699, -------------- Report Sync ------------
Event: time 1710070752.012602, type 3 (Absolute), code 57 (MT Tracking ID), value 0
Event: time 1710070752.012602, type 3 (Absolute), code 53 (MT X), value 610
Event: time 1710070752.012602, type 3 (Absolute), code 54 (MT Y), value 2230
Event: time 1710070752.012602, type 3 (Absolute), code 58 (Pressure on contact area), value 89
Event: time 1710070752.012602, type 3 (Absolute), code 48 (MT Major Axis Ellipse), value 95
Event: time 1710070752.012602, -------------- Report Sync ------------
^C
[kterm]# 
[kterm]# evtest /dev/input/touch [J[J[J[J[J[J[Ji[J
by-path/  event1    event3    event5    touch
event0    event2    event4    stylus    uinput

[kterm]# evtest /dev/input/[J
by-path/  event1    event3    event5    touch
event0    event2    event4    stylus    uinput

[kterm]# evtest /dev/input/[Jui
[kterm]# evtest /dev/input/uinput [J
evtest: can't get version: Invalid argument
[kterm]# exit
Script done on Sun 10 Mar 2024 02:39:39 PM GMT+3:0100
```

The stock `evtest` is slightly more annoying, but that'll do (I've seen much worse/older imps ;p).

**Comment by mergen3107** [comment]
@NiLuJe 
Any objections of implementing this fix for now? https://github.com/koreader/koreader/issues/11392#issuecomment-1924558908
I keep getting user's complains with Scribe who cannot run KOReader just because of this error.

Otherwise, maybe we can guard it? Like `try` getting the file, then `catch` it, but set `hasGsensor` to false and let KOReader start at least.

**Comment by Frenzie** [comment]
Lua has pcall, but simply that `if std_out` seems fine by me. Please submit a PR.

**Comment by NiLuJe** [comment]
IIRC, there's already something like that being done on the Oasis, so, sure, it's better than nothing, as it does otherwise cause a hard crash on startup ;).

**Comment by mergen3107** [comment]
> there's already something like that being done on the Oasis

Yes, this is where I got the inspiration from :D 
Thanks! Will PR soon

**Comment by mergen3107** [comment]
How can I access `hasGsensor` in the init?
I tried:
```
    -- Get accelerometer device
    logger.dbg("hasGsensor", KindleScribe.hasGsensor)
```
but it says nil. I also tried `self.hasGsensor` and just `hasGsensor`.

**Comment by mergen3107** [comment]
Idea is to modify this to `no` if no accelerometer device is found.

**Comment by Frenzie** [comment]
Which line number (give or take)? You can also link to a specific line by clicking on the line number, → … → copy permalink.

**Comment by mergen3107** [comment]
@Frenzie 
https://github.com/koreader/koreader/blob/1b893d43c9cfe6a46a5a9930bd679a578d7d7a23/frontend/device/kindle/device.lua#L668

**Comment by Frenzie** [comment]
I meant where you want to access it. :-) Presumably it's something like self.hasGSensor.

**Comment by mergen3107** [comment]
> Presumably it's something like self.hasGSensor

This yielded `nil` :D

**Comment by Frenzie** [comment]
Ah, that's too late. You'd have to do it before Kindle.init(self) to relevantly affect the initialization code.

Perhaps you can just make it:

```lua
    hasGSensor = io.popen("grep -A4 'accel' /proc/bus/input/devices | grep -o 'event[0-9]'", "r") and yes or no,
```

Over on line 669.

Unless @NiLuJe finds that improper of course.

**Comment by mergen3107** [comment]
I don't think that check is enough. 
This file can still report that, say, "event1" is the gyro. But we also need to check that /dev/input/event1 exists indeed

**Comment by Frenzie** [comment]
In that case the whole function, but you can't do the self.input.open stuff there anyway. Something along these lines:

```
hasGSensor = function()
     local std_out = io.popen("grep -A4 'accel' /proc/bus/input/devices | grep -o 'event[0-9]'", "r")
     if std_out then
         local gyro_dev = std_out:read("*line")
         std_out:close()
         logger.dbg("gyro_dev", gyro_dev)
         if gyro_dev then
             return yes
         end
     end
     return no
end,
```

**Comment by NiLuJe** [comment]
> > Presumably it's something like self.hasGSensor
> 
> This yielded `nil` :D

Then you're doing something wrong ;).

In `WhicheverDevice:init`, `self` points to `WhicherverDevice`, so everything should track.

c.f., Kobo's init, which is far more guresome and does do a lot of stuff like that ;).

**Comment by NiLuJe** [comment]
> Ah, that's too late. You'd have to do it before Kindle.init(self) to relevantly affect the initialization code.

Oh, that, too, yeah ;).

**Comment by NiLuJe** [comment]
@mergen3107: Can you run the following binary (`input_scan`) on your Scribe and then post the terminal output?

I've been working on the input detection/classification thing I mentioned a while ago, and the Scribe's stylus is potentially going to be finicky, so I'd like to check whether I got it right or not...

[InputScan-v1.25.0-393-g2775e6f.tar.gz](https://github.com/koreader/koreader/files/15285150/InputScan-v1.25.0-393-g2775e6f.tar.gz)

**Comment by mergen3107** [comment]
@NiLuJe 
Here you go!
```
[root@kindle koreader]# ./input_scan
[FBInk] /dev/input/event0: `bd71828-pwrkey` = KEY | POWER_BUTTON
[FBInk] /dev/input/event1: `kx132-accel` = ACCELEROMETER
[FBInk] /dev/input/event2: `WacomDigitizer` = TABLET
[FBInk] /dev/input/event3: `pt_mt` = TOUCHSCREEN
[FBInk] /dev/input/event4: `stylus-custom` = TABLET | SCALED_TABLET
```

**Comment by mergen3107** [comment]
FWIW, currently I have to use:
1) event3 as "touch" as the /dev/input/touch symlink
2) event4 as "pen"
3) event1 as "accelerometer"

There is also "stylus" symlink (/dev/input/stylus -> event2), but it didn't work as pen input.

**Comment by mergen3107** [comment]
Would you like me to ask people over 4PDA to run the script as well?

**Comment by NiLuJe** [comment]
Oops, I fumbled the previous build, it was using the Kobo events instead of the Kindle ones ;). On the upside, the pen stuff worked out, so there's that ;p.

Can you try this one?

[InputScan-v1.25.0-393-g2775e6f-KINDLE.tar.gz](https://github.com/koreader/koreader/files/15285420/InputScan-v1.25.0-393-g2775e6f-KINDLE.tar.gz)

Ideally, `kx132-accel` should gain a `ROTATION_EVENT` tag.

----

I wouldn't mind a pass on the other Scribes with a `bma4xy` accelerometer, just in case, yeah ;).

**Comment by NiLuJe** [comment]
Hmm, and given the `evtest` results, the output of `./input_scan -m rotation -x tablet` might be helpful. (Assuming you did get a bunch of new `ROTATION_EVENT` flags with the proper binaries ;p).

**Comment by mergen3107** [comment]
Here it is:
```
[root@kindle koreader]# ./input_scan2_kindle
[FBInk] /dev/input/event0: `bd71828-pwrkey` = KEY | POWER_BUTTON
[FBInk] /dev/input/event1: `kx132-accel` = ACCELEROMETER | ROTATION_EVENT
[FBInk] /dev/input/event2: `WacomDigitizer` = TABLET | ROTATION_EVENT
[FBInk] /dev/input/event3: `pt_mt` = TOUCHSCREEN
[FBInk] /dev/input/event4: `stylus-custom` = TABLET | ROTATION_EVENT | SCALED_TABLET
```

**Comment by mergen3107** [comment]
I asked people over 4PDA, waiting for replies

**Comment by mergen3107** [comment]
> ./input_scan -m rotation -x tablet

```
[root@kindle koreader]# ./input_scan2_kindle -m rotation -x tablet
[InputScan] Requested match mask (any): 0x00800000
[InputScan] Requested exclusion mask (any): 0x00000020
[test_key] [bd71828-pwrkey] checking bit block 0 for any keys; found=0
[test_key] [bd71828-pwrkey] checking bit block 32 for any keys; found=0
[test_key] [bd71828-pwrkey] checking bit block 64 for any keys; found=0
[test_key] [bd71828-pwrkey] checking bit block 96 for any keys; found=1
[test_key] [bd71828-pwrkey] checking bit block 128 for any keys; found=1
[test_key] [bd71828-pwrkey] checking bit block 160 for any keys; found=1
[test_key] [bd71828-pwrkey] checking bit block 192 for any keys; found=1
[test_key] [bd71828-pwrkey] checking bit block 224 for any keys; found=1
[FBInk] /dev/input/event0: `bd71828-pwrkey` = KEY | POWER_BUTTON
[test_key] [kx132-accel] no EV_KEY capability
[FBInk] /dev/input/event1: `kx132-accel` = ACCELEROMETER | ROTATION_EVENT
[test_key] [WacomDigitizer] checking bit block 0 for any keys; found=0
[test_key] [WacomDigitizer] checking bit block 32 for any keys; found=0
[test_key] [WacomDigitizer] checking bit block 64 for any keys; found=0
[test_key] [WacomDigitizer] checking bit block 96 for any keys; found=0
[test_key] [WacomDigitizer] checking bit block 128 for any keys; found=0
[test_key] [WacomDigitizer] checking bit block 160 for any keys; found=0
[test_key] [WacomDigitizer] checking bit block 192 for any keys; found=0
[test_key] [WacomDigitizer] checking bit block 224 for any keys; found=0
[FBInk] /dev/input/event2: `WacomDigitizer` = TABLET | ROTATION_EVENT
[test_key] [pt_mt] checking bit block 0 for any keys; found=0
[test_key] [pt_mt] checking bit block 32 for any keys; found=0
[test_key] [pt_mt] checking bit block 64 for any keys; found=0
[test_key] [pt_mt] checking bit block 96 for any keys; found=0
[test_key] [pt_mt] checking bit block 128 for any keys; found=0
[test_key] [pt_mt] checking bit block 160 for any keys; found=0
[test_key] [pt_mt] checking bit block 192 for any keys; found=0
[test_key] [pt_mt] checking bit block 224 for any keys; found=0
[FBInk] /dev/input/event3: `pt_mt` = TOUCHSCREEN
[test_key] [stylus-custom] checking bit block 0 for any keys; found=0
[test_key] [stylus-custom] checking bit block 32 for any keys; found=0
[test_key] [stylus-custom] checking bit block 64 for any keys; found=0
[test_key] [stylus-custom] checking bit block 96 for any keys; found=0
[test_key] [stylus-custom] checking bit block 128 for any keys; found=0
[test_key] [stylus-custom] checking bit block 160 for any keys; found=0
[test_key] [stylus-custom] checking bit block 192 for any keys; found=0
[test_key] [stylus-custom] checking bit block 224 for any keys; found=0
[FBInk] /dev/input/event4: `stylus-custom` = TABLET | ROTATION_EVENT | SCALED_TABLET
[InputScan] Found 5 readable input devices
[InputScan] Device bd71828-pwrkey @ /dev/input/event0 is classified as 0x00010040 (matched: 0                                                                                                                      )
[InputScan] Device kx132-accel @ /dev/input/event1 is classified as 0x00800100 (matched: 1)
[InputScan] Device WacomDigitizer @ /dev/input/event2 is classified as 0x00800020 (matched: 0                                                                                                                      )
[InputScan] Device pt_mt @ /dev/input/event3 is classified as 0x00000008 (matched: 0)
[InputScan] Device stylus-custom @ /dev/input/event4 is classified as 0x01800020 (matched: 0)

```

**Comment by mergen3107** [comment]
```
[InputScan] Device bd71828-pwrkey @ /dev/input/event0 is classified as 0x00010040 (matched: 0
```
There are some missing parentheses in the output - I copied it as is from PuTTy

**Comment by dangerweenie** [comment]
I can't launch KOreader on Kindle Scribe either - the screen just flashes and goes back to home (tried clean install and nightly build)

./koreader/crash.log:

`/var/tmp/koreader.sh: line 326: ./reader.lua: not found
/var/tmp/koreader.sh: line 326: ./reader.lua: not found`

**Comment by mergen3107** [comment]
@dangerweenie 
You need kindlehf version of noghtly
